sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -0,0 +1,1637 @@
+----
+
+|Invokes the execution of a graph. Only one instance of `graph` may be executing,
+or pending execution, at any time. Concurrent graph execution can be achieved by
+finalizing a graph in modifiable state into multiple graphs in executable state.
+
+Parameters:
+
+* `graph` - Graph object to execute.
+
+Exceptions:
+
+* Throws synchronously with error code `invalid` if the handler is submitted
+  to a queue which doesn't have a SYCL context which matches the context of
+  the executable graph.
+
+* Throws synchronously with error code `invalid` if a previous submission of
+  `graph` has yet to complete execution.
+|===
+
+=== Thread Safety
+
+The new functions in this extension are thread-safe, the same as member
+functions of classes in the base SYCL specification. If user code does
+not perform synchronization between two threads accessing the same queue,
+there is no strong ordering between events on that queue, and the kernel
+submissions, recording and finalization will happen in an undefined order.
+
+When one thread ends recording on a queue while another
+thread is submitting work, which kernels will be part of the subsequent
+graph is undefined. If user code enforces a total order on the queue
+events, then the behavior is well-defined, and will match the observable
+total order.
+
+The returned value from the `queue::ext_oneapi_get_state()` should be
+considered immediately stale in multi-threaded usage, as another thread could
+have preemptively changed the state of the queue.
+
+=== Exception Safety
+
+In addition to the destruction semantics provided by the SYCL
+{crs}[common reference semantics], when the last copy of a modifiable
+`command_graph` is destroyed recording is ended on any queues that are recording
+to that graph, equivalent to `+this->end_recording()+`.
+
+As a result, users don't need to manually wrap queue recording code in a
+`try` / `catch` block to reset the state of recording queues on an exception
+back to the executing state. Instead, an uncaught exception destroying the
+modifiable graph will perform this action, useful in RAII pattern usage.
+
+=== Storage Lifetimes [[storage-lifetimes]]
+
+The lifetime of any buffer recorded as part of a submission
+to a command graph will be extended in keeping with the common reference
+semantics and buffer synchronization rules in the SYCL specification. It will be
+extended either for the lifetime of the graph (including both modifiable graphs
+and the executable graphs created from them) or until the buffer is no longer
+required by the graph (such as after being replaced through executable graph update).
+
+If a buffer created with a host data pointer is recorded as part of a submission to
+a command graph, the lifetime of that host data will also be extended by taking a
+copy of that data inside the buffer. This copy will occur during the call to
+`queue::submit()`. Users can opt-out of this behavior by passing the property
```

Comment:
I agree that this could probably be left as an implementation detail so we could remove that specific description of when it will occur. This information could be in our implementation design docs instead.

We could potentially do the copy in the buffer destructor, though that has the usual concerns of causing a termination if something goes wrong with the copy and we have to throw. Although this is a simplified example you've provided, this is the kind of situation where the `no_host_copy` property is useful. I think that taking a copy unless the user explicitly says it's OK not to is much safer which is desirable, even if perhaps in some edge cases where the user doesn't know to use the property (even though they could), it is less performant. It's worth noting also that these copies would happen during recording where performance is less critical, and would only happen on the first instance of a buffer being used this way, which I think is a fair trade off.

Reply:
Yes, I see your point about it being harder to do the copy in the buffer destructor.  It seems like we agree that the spec shouldn't dictate when the copy happens, so that leaves room for you to change the implementation later if it seems beneficial.